skimage/restoration/_nl_means_denoising.pyx

```diff
@@ -0,0 +1,596 @@
+            y_start = y - offset
+            y_end = y + offset + 1
+            # Coordinates of test pixel and patch bounds
+            for i in range(max(- d, offset - x),
+                           min(d + 1, n_x - x - 1)):
+                x_start_i = x_start + i
+                x_end_i = x_end + i
+                for j in range(max(- d, offset - y),
+                               min(d + 1, n_y - y - 1)):
+                    y_start_j = y_start + j
+                    y_end_j = y_end + j
+                    weight = patch_distance_2drgb(
+                                padded[x_start: x_end,
+                                       y_start: y_end, :],
+                                padded[x_start_i: x_end_i,
+                                       y_start_j: y_end_j, :],
+                                w, s)
+                    weight_sum += weight
+                    for color in range(3):
+                        new_values[color] += weight * padded[x + i, y + j,
+                                                             color]
+            for color in range(3):
+                result[x, y, color] = new_values[color] / weight_sum
+    return result[offset:-offset, offset:-offset]
+
+
+@cython.cdivision(True)
+@cython.boundscheck(False)
+def _nl_means_denoising_3d(image, int s=7,
+            int d=13, float h=0.1):
+    """
+    Perform non-local means denoising on 3-D array
+
+    Parameters
+    ----------
+    image: ndarray
+        input data to be denoised
+
+    s: int, optional
+        size of patches used for denoising
+
+    d: int, optional
+        maximal distance in pixels where to search patches used for denoising
+
+    h: float, optional
+        cut-off distance (in gray levels)
+    """
+    if s % 2 == 0:
+        s += 1  # odd value for symmetric patch
+    cdef int n_x, n_y, n_z
+    n_x, n_y, n_z = image.shape
+    cdef int offset = s / 2
+    # padd the image so that boundaries are denoised as well
+    cdef DTYPE_t [:, :, ::1] padded = np.ascontiguousarray(util.pad(
+                                        image.astype(np.float32),
+                                        offset, mode='reflect'))
+    cdef DTYPE_t [:, :, ::1] result = padded.copy()
+    cdef float A = ((s - 1.) / 4.)
+    cdef float new_value
+    cdef float weight_sum, weight
+    xg, yg, zg = np.mgrid[-offset: offset + 1, -offset: offset + 1,
+                            -offset: offset + 1]
+    cdef DTYPE_t [:, :, ::1] w = np.ascontiguousarray(np.exp(
+                            - (xg ** 2 + yg ** 2 + zg ** 2) / (2 * A ** 2)).
+                                astype(np.float32))
+    cdef float distance
+    cdef int x, y, z, i, j, k
+    cdef int x_start, x_end, y_start, y_end, z_start, z_end
+    cdef int x_start_i, x_end_i, y_start_j, y_end_j, z_start_k, z_end_k
+    w = 1. / (np.sum(w) * h ** 2) * w
+    # Coordinates of central pixel and patch bounds
+    for x in range(offset, n_x + offset):
```

Comment:
This *might* be doable with `np.nditer`. I don't know how much worse that would be for performance, but it would certainly help readability!